external-crates/move/crates/move-regex-borrow-graph/src/references.rs

```diff
@@ -0,0 +1,439 @@
+// joining
+//**************************************************************************************************
+
+impl<Loc: Copy, Lbl: Ord + Clone> Node<Loc, Lbl> {
+    // adds all edges in other to self, where the successor/predecessor is in mask
+    pub(crate) fn join(&mut self, mask: &BTreeSet<Ref>, other: &Self) -> usize {
+        let mut size_increase = 0usize;
+        for (s, edge) in other.successors.iter().filter(|(s, _)| mask.contains(s)) {
+            for LocRegex { loc, regex } in &edge.regexes {
+                size_increase =
+                    size_increase.saturating_add(self.add_regex(*loc, regex.clone(), *s));
+            }
+        }
+        for p in other.predecessors.iter().filter(|p| mask.contains(p)) {
+            self.add_predecessor(*p);
+        }
+
+        size_increase
+    }
+}
+
+//**************************************************************************************************
+// invariants
+//**************************************************************************************************
+
+impl Ref {
+    pub fn is_canonical(&self) -> bool {
+        matches!(self.0, Ref_::Canonical(_))
+    }
+
+    pub fn is_fresh(&self) -> bool {
+        matches!(self.0, Ref_::Fresh(_))
+    }
+}
+
+impl<Loc, Lbl: Ord> Edge<Loc, Lbl> {
+    pub(crate) fn check_invariant(&self) {
+        #[cfg(debug_assertions)]
+        {
+            let mut calculated_size = 0;
+            for r in &self.regexes {
+                calculated_size += r.regex.abstract_size();
+            }
+            debug_assert_eq!(calculated_size, self.abstract_size);
+            debug_assert!(!self.regexes.is_empty());
+        }
+    }
+}
+
+impl<Loc, Lbl: Ord> Node<Loc, Lbl> {
+    pub(crate) fn check_invariant(&self) {
+        #[cfg(debug_assertions)]
+        {
+            let is_canonical = self.ref_.is_canonical();
+            for s in self.successors.keys() {
+                debug_assert_eq!(is_canonical, s.is_canonical());
+            }
+            for p in &self.predecessors {
+                debug_assert_eq!(is_canonical, p.is_canonical());
+            }
+            debug_assert!(self.self_epsilon.is_epsilon());
+            debug_assert!(!self.successors.contains_key(&self.ref_));
+            debug_assert!(!self.predecessors.contains(&self.ref_));
+            let mut calculated_size = 1;
+            for edge in self.successors.values() {
+                edge.check_invariant();
+                calculated_size += edge.abstract_size();
+            }
+            debug_assert_eq!(calculated_size, self.abstract_size);
+        }
+    }
+}
+
+//**************************************************************************************************
+// traits
+//**************************************************************************************************
+
+impl<Loc, Lbl: PartialEq> PartialEq for LocRegex<Loc, Lbl> {
```

Comment:
Is there a risk of two LocRegexes with the same label but different locs clobbering each other in the regex map? Is that a problem when it happens?